### PR TITLE
Added a sample Kerberos group authentication external_acl helper

### PR DESCRIPTION
**Pull request by huaraz**

### CODE REVIEW

File: src/auth/negotiate/kerberos/negotiate_external_acl [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#
+kinit -kt $KTNAME $PRINCIPAL
+SID=`ldapsearch -LLL -Ygssapi -H ldap://$DC:389 -s sub -b "BASEDN" "(CN=$GROUPNAME)" objectsid 2>&1 | awk '{ if ( $0 ~/^object/ ) print $2}'`
+
+(>&2 echo "`date +"%Y/%m/%d %H:%M:%S"`| negotiate_external_acl: SID=$SID")
+
+#
+# Read from STDIN as a series of lines into 'input'."
+#
+while read input
```
Comment (rousskov) [inline src/auth/negotiate/kerberos/negotiate_external_acl]: FYI: This is is the line where Semaphore CI found a trailing space.

File: src/auth/negotiate/kerberos/negotiate_external_acl [hunk 2]
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
+#
+# This external acl helper checks the Kerberos group information without
+# continues ldap queries by leveraging the kerbors ticket group information
+# i.e. PAC field
+#
+# It requires only during startup aldap query to get the SID for the group name.
```
Comment (rousskov) [inline src/auth/negotiate/kerberos/negotiate_external_acl]: I would rephrase for clarity. For example:

```
During startup, it sends an ldapsearch query to get the SID for the group name.
```

**Comment by rousskov** [comment]
Removed `I updated the script based on feedback` comment from the PR description so that it does not become a commit message body.

**Comment by rousskov** [comment]
@eduard-bagdasaryan, is the new 60 second Jenkins delay operational? If yes, it looks like it did not help this PR. Jenkins is still quitting with `reference is not a tree` [errors](https://build.squid-cache.org/job/5-pr-test/505/compiler=clang,label=d-debian-stretch/console).

**Comment by rousskov** [comment]
@huaraz, thank you for addressing most of my feedback. I hope somebody more qualified in Kerberos issues will review this soon.

**Comment by rousskov** [comment]
I see `+ sleep 60` in the [console](https://build.squid-cache.org/job/5-pr-test/505/compiler=clang,label=d-debian-stretch/console) output. I assume it is sequential and should delay repo cloning/checkout. Why is it not helping?

**Comment by rousskov** [comment]
> Why is it not helping?

For the record, @eduard-bagdasaryan have identified the cause of these build errors as a known and unresolved Jenkins ghprb-plugin design [problem](https://github.com/jenkinsci/ghprb-plugin/issues/489). Sleeping does not solve this problem. We are now working on a workaround that would identify and retry build tests that failed due to this problem.

File: src/auth/negotiate/kerberos/Makefile.am [hunk 3]
```diff
@@ -10,7 +10,8 @@ include $(top_srcdir)/src/Common.am
 EXTRA_DIST= \
 	README \
 	required.m4 \
-	negotiate_kerberos_auth.8
+	negotiate_kerberos_auth.8 \
+	negotiate_external_acl
```
Comment (yadij) [inline src/auth/negotiate/kerberos/Makefile.am]: external ACL helpers go under src/acl/external/FOO/ now with ext_FOO_acl naming.

File: src/auth/negotiate/kerberos/negotiate_external_acl [hunk 4]
```diff
@@ -0,0 +1,63 @@
+#!/bin/sh
+#
```
Comment (yadij) [inline src/auth/negotiate/kerberos/negotiate_external_acl]: missing copyright blurb.

File: src/auth/negotiate/kerberos/negotiate_external_acl (hunk 2 above)
Comment (yadij) [reply src/auth/negotiate/kerberos/negotiate_external_acl]: That is almost a conjunctive sentence; "During startup, send an LDAP ..." or "Sends an LDAP query during startup to get ..."

File: src/auth/negotiate/kerberos/Makefile.am (hunk 3 above)
Comment (huaraz) [reply src/auth/negotiate/kerberos/Makefile.am]: OK It was intended as a simple example to show how the Kerberos PAC field extracted by the auth helper can be efficiently used.  If you agree I can move it to src/acl/external/negotiate_kerberos_group

**Comment by rousskov** [comment]
> We are now working on a workaround that would identify and retry build tests that failed due to this problem.

@eduard-bagdasaryan has found and deployed a better workaround. With two additional plugins ([Text-finder](https://wiki.jenkins.io/display/JENKINS/Text-finder+Plugin) and [TextFinderRunConditionPlugin](https://wiki.jenkins.io/display/JENKINS/Text+Finder+Run+Condition+Plugin)), Jenkins should now retry builds that failed with a `fatal: reference is not a tree` string in at least one of the node console logs.

File: src/auth/negotiate/kerberos/ext_negotiate_kerberos_acl [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+#    the Libraries in the source distribution.
+#
+#############################################################################
+# 
+# This external acl helper checks the Kerberos group information without
+# continues ldap queries by leveraging the kerbors ticket group information
+# i.e. PAC field. During startup the helper determines the objectsid for the
+# group and compares it to the kerberos token group information. This must
+# be a Security group as these are only stored in the Kerberos ticket.
+#
```
Comment (rousskov) [inline src/auth/negotiate/kerberos/ext_negotiate_kerberos_acl]: This comment is attached to one of those lines with trailing whitespace.

**Comment by huaraz** [comment]
Hi Amos,

what acl helper name do you suggest ?  i.e. move it to src/acl/external/negotiate_kerberos_group ?

Markus

**Comment by yadij** [comment]
Going by the schema we have this would be ext_kerberos_ldap_group_acl. Since there is already a helper for that would it be easy enough to extend that helper?

If not maybe we should start a category "group_sid" or "sid".

**Comment by huaraz** [comment]
The main difference between ext_kerberos_ldap_group_acl  and this helper is, that ext_kerberos_ldap_group_acl  uses "just" Kerberos to authenticate to the ldap server using the squid keytab ( otherwise it is the same as the LDAP_group helper with some automagic ) and this helper uses the Group information from the Kerberos ticket i.e. much more efficient in a pure AD environment.

**Comment by huaraz** [comment]
So I think ext_kerberos_sid_group_acl might be the best option.

**Comment by yadij** [comment]
That seems okay to me. Its at least consistent with the other kerberos helpers, even if LDAP dependency is not made clear by the name.

**Comment by huaraz** [comment]
I have now moved the sample shell script to its own external helper directory as a perl script.

File: src/auth/negotiate/kerberos/Makefile.am [hunk 6]
```diff
@@ -10,7 +10,8 @@ include $(top_srcdir)/src/Common.am
 EXTRA_DIST= \
 	README \
 	required.m4 \
-	negotiate_kerberos_auth.8
+	negotiate_kerberos_auth.8 \
+	ext_negotiate_kerberos_acl
```
Comment (yadij) [inline src/auth/negotiate/kerberos/Makefile.am]: reference to obsolete helper name. needs removing to fix the build errors. This file should now not have any changes.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                	$ans = &check($group);
+                	last if $ans eq "OK";
+        	}
+        	&debug("DEBUG: Sending $ans to squid");
+        	print "$ans\n";
+        } else {
+        	&debug("DEBUG: Sending ERR to squid");
+        	print "ERR\n";
+	}
+}
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: our tools unfortunately need an empty line at the end of files. please add one here.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        print "\t-D domain controller (required)\n";
+        print "\t-b base dn (required)\n";
+        print "\t-G list of AD group names (seperated by : ) (required)\n";
+        exit;
+}
+
+init();
+&debug("INFO: Debugging mode ON.");
+
+
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: double empty line. please remove.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+# Command line options processing
+#
+sub init()
+{
+	use Getopt::Std;
+	my $errmsg;
+	my $opt_string = 'hdD:p:b:G:';
+	getopts( "$opt_string", \%opt ) or usage();
+	usage() if $opt{h};
+	usage() if not defined $opt{D};
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: Just call:
 Pod::Usage::pod2usage(1) if not defined $opt{D};

same for other calls to usage()

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+		}
+	}
+	&info("ERROR: Exit as no sid was found for any group") if ! @ADgroupSIDs;
+	exit 99 if ! @ADgroupSIDs;
+}
+
+#
+# Message about this program and how to use it
+#
+sub usage()
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: you are pulling in Pod::Usage which generates the help output from documentation SYNOPSIS section. No need to define a custom usage() function.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+=head1 REPORTING BUGS
+
+Bug reports need to be made in English.
+See http://wiki.squid-cache.org/SquidFaq/BugReporting for details of what you need to     include with your bug report.
+
+Report bugs or bug fixes using http://bugs.squid-cache.org/
+
+Report serious security bugs to I<Squid Bugs <squid-bugs@lists.squid-cache.org>>
+
+Report ideas for new improvements to the I<Squid Developers mailing list <squid-dev@li    sts.squid-cache.org>>
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: broken whitespace in the email contact. same for some lines above.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+
+=item B<-G AD group name>
+
+AD group name to be used for SID lookup. List seperated by :
+
+=back
+
+=head1 CONFIGURATION
+
+  auth_param negotiate program /path/to/negotiate_wrapper_auth -d --ntlm /path/to/ntlm_auth --helper-protocol=squid-2.5-ntlmssp --domain example.com --kerberos /path/to/negotiate_kerberos_auth -d -s GSS_C_NO_NAME -k /path/to/squid.keytab -t none
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: FYI: squid.conf supports lien wrapping if the line ends with as slash '\' and the next starts with whitespace. So you can prevent line wrapping issues by wrapping example lines to be shorter than 80 characters.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+=cut
+
+#
+# Version history:
+#   2018-06-10 Markus Moeller <markus_moeller@compuserve.com>
+#               Initial release
+#
+# Globals
+#
+use File::Basename;
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: Please put all the module dependencies together at the top so packagers can easily identify whats needed. (I know a lot of squid existing helpers don't do that - its a bug still to be fixed.)

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+use Pod::Usage;
+use Getopt::Long;
+
+=pod
+
+=head1 NAME
+
+ ext_kerberos_sid_group_acl - external ACL helper for Squid to verify AD Domain group membership using sid.
+
+
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: double empty line. please remove.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+Domain controller to contact to lookup group SID
+
+=item B<-b base DN>
+
+Base DN for ldap search
+
+=item B<-G AD group name>
+
+AD group name to be used for SID lookup. List seperated by :
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: spelling: "separated".

Also, it is clearer to write " by a colon (:)" to clarify that the colon is the delimiter, not punctuation of the documentation sentence.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+
+  auth_param negotiate program /path/to/negotiate_wrapper_auth -d --ntlm /path/to/ntlm_auth --helper-protocol=squid-2.5-ntlmssp --domain example.com --kerberos /path/to/negotiate_kerberos_auth -d -s GSS_C_NO_NAME -k /path/to/squid.keytab -t none
+  external_acl_type sid_check ipv4 %LOGIN %note{group} /path/to/kerberos_sid_group_acl -p principal -D dc1.example.com -b "DC=example,DC=com" -G Group1:Group2
+  acl squid_allow external sid_check
+  acl allowed_group external sid_check 
+  http_access allow allowed_group
+
+If the local perl interpreter is in a unusual location it may need to be added:
+
+  external_acl_type sid_check ipv4 %LOGIN %note{group} /path/to/perl /path/to/kerberos_sid_group_acl -p principal -D dc1.example.com -b "DC=example,DC=com" -G Group1:Group2
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: Do we really need to force IPv4-only contact here?

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+See http://wiki.squid-cache.org/SquidFaq/BugReporting for details of what you need to     include with your bug report.
+
+Report bugs or bug fixes using http://bugs.squid-cache.org/
+
+Report serious security bugs to I<Squid Bugs <squid-bugs@lists.squid-cache.org>>
+
+Report ideas for new improvements to the I<Squid Developers mailing list <squid-dev@li    sts.squid-cache.org>>
+
+=head1 SEE ALSO
+
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: Since this helper is described earlier as requiring use alongside the negotiate_kerberos_auth helper, there should probably be a reference from here to:  negotiate_kerberos_auth(8)

Comment (huaraz) [reply src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: Thank you 
Markus

**Comment by huaraz** [comment]
Not sure where the spaces came from.  I am not sure if the sample config lines are now as you proposed.

File: src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in [hunk 18]
```diff
@@ -59,7 +59,7 @@ AD group name to be used for SID lookup. List separated by a colon (:)
 
 =head1 CONFIGURATION
 
-  auth_param negotiate program /path/to/negotiate_wrapper_auth -d 
+  auth_param negotiate program /path/to/negotiate_wrapper_auth -d
```
Comment (yadij) [inline src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: these are missing the '\' at end of lines to indicate wrapping.

Comment (yadij) [reply src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: hmm, github does not seem to like showing slashes in comments.

Comment (huaraz) [reply src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: Sorry I don't follow. Is there something I should change ?

Comment (rousskov) [reply src/acl/external/kerberos_sid_group/ext_kerberos_sid_group_acl.pl.in]: I believe @yadij is implying that you should add backslashes (`\`) at end of this and probably other similar lines to indicate command/line wrapping.

**Comment by huaraz** [comment]
Is there anything else I should adjust ?

Thank you
Markus

**Comment by rousskov** [comment]
>  squid-anubis added `M-waiting-staging-checks` and removed `M-waiting-staging-checks` labels [8 minutes ago](https://github.com/squid-cache/squid/pull/202#event-1699269652)

@eduard-bagdasaryan, please check what happened to Anubis-controlled labels in this PR. The Semaphore CI check is currently pending, but GitHub shows no `M-waiting-staging-checks` label. Thus, I assume that Anubis added and then quickly (and incorrectly?) removed the `M-waiting-staging-checks` label.

**Comment by eduard-bagdasaryan** [comment]
> please check what happened to Anubis-controlled labels 

The bot got stuck with this PR as you can see: there are lots of recreated failed staged commits. The bot incorrectly compared empty PR descriptions (this PR has the only title). I hope it should be fixed now.

**Comment by eduard-bagdasaryan** [comment]
> Jenkins should now retry builds that failed with a `fatal: reference is not a tree` string in at least one of the node console logs

I increased the number of retries from 1 to  5.
